registry/registry.go

```diff
@@ -0,0 +1,360 @@
+package registry
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/Sirupsen/logrus/formatters/logstash"
+	"github.com/bugsnag/bugsnag-go"
+	"github.com/docker/distribution/configuration"
+	ctxu "github.com/docker/distribution/context"
+	"github.com/docker/distribution/health"
+	"github.com/docker/distribution/registry/handlers"
+	"github.com/docker/distribution/registry/listener"
+	"github.com/docker/distribution/uuid"
+	"github.com/docker/distribution/version"
+	gorhandlers "github.com/gorilla/handlers"
+	"github.com/spf13/cobra"
+	"github.com/yvasiyarov/gorelic"
+	"golang.org/x/net/context"
+)
+
+// Cmd is a cobra command for running the registry.
+var Cmd = &cobra.Command{
+	Use:   "registry <config>",
+	Short: "registry stores and distributes Docker images",
+	Long:  "registry stores and distributes Docker images.",
+	Run: func(cmd *cobra.Command, args []string) {
+		if showVersion {
+			version.PrintVersion()
+			return
+		}
+
+		config, err := resolveConfiguration(args)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "configuration error: %v\n", err)
+			cmd.Usage()
+			os.Exit(1)
+		}
+
+		registry, err := NewRegistry(context.Background(), config)
+		if err != nil {
+			log.Fatalln(err)
+		}
+
+		err = registry.Serve()
+		if err != nil {
+			log.Fatalln(err)
+		}
+	},
+}
+
+var showVersion bool
+
+func init() {
+	Cmd.PersistentFlags().BoolVarP(&showVersion, "version", "v", false, "show the version and exit")
+}
+
+// A Registry represents a complete instance of the registry.
+type Registry struct {
```

Comment:
I somewhat agree but think it is outside the scope of the goal of this particular PR. This refactor is going to have be multi-step.

@aaronlehmann Do we want a TODO here?